File: src/main/java/com/contentvelocity/service/CrawlerService.java
package com.contentvelocity.service;

import com.contentvelocity.model.PageData;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.net.URI;
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Service
public class CrawlerService {

    private static final Logger log = Logger.getLogger(CrawlerService.class.getName());

    private final ObjectMapper schemaMapper = new ObjectMapper();

    private static final Map<String, Integer> SCHEMA_EXPECTED_FIELDS = Map.of(
        "Article", 8, "BlogPosting", 8, "NewsArticle", 8,
        "Product", 9, "FAQPage", 4, "WebSite", 5,
        "WebPage", 5, "Organization", 6, "LocalBusiness", 7,
        "Person", 5
    );

    private static final Pattern FACTUAL_PATTERN = Pattern.compile(
        "\\d+([,.]\\d+)?(%|\\s*(percent|million|billion|thousand))?|" +
        "\\b(according to|research|study|studies|report|survey|data|statistics|found that)\\b",
        Pattern.CASE_INSENSITIVE
    );

    private static final Pattern TECHNICAL_TERM_PATTERN = Pattern.compile(
        "\\b[A-Z][a-zA-Z0-9]*(?:[\\s-][A-Z][a-zA-Z0-9]*){0,2}\\b"
    );

    @Value("${scanner.connect-timeout-ms:10000}")
    private int connectTimeout;

    @Value("${scanner.read-timeout-ms:15000}")
    private int readTimeout;

    @Value("${scanner.user-agent}")
    private String userAgent;

    public PageData crawl(String url) throws Exception {
        PageData data = new PageData();
        data.url = url;

        try {
            URI uri = new URI(url);
            data.slug = uri.getPath();
            data.baseDomain = uri.getHost();
        } catch (Exception e) {
            data.slug = url;
        }

        Document doc = Jsoup.connect(url)
                .userAgent(userAgent)
                .timeout(connectTimeout)
                .followRedirects(true)
                .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8")
                .header("Accept-Language", "en-US,en;q=0.9")
                .header("Accept-Encoding", "gzip, deflate, br")
                .header("Cache-Control", "no-cache")
                .header("Pragma", "no-cache")
                .header("Upgrade-Insecure-Requests", "1")
                .header("Sec-Fetch-Dest", "document")
                .header("Sec-Fetch-Mode", "navigate")
                .header("Sec-Fetch-Site", "none")
                .header("Sec-Fetch-User", "?1")
                .ignoreHttpErrors(false)
                .get();

        // Title
        data.title = doc.title();

        // Meta description
        Element metaDesc = doc.selectFirst("meta[name=description]");
        if (metaDesc == null) metaDesc = doc.selectFirst("meta[property=og:description]");
        data.metaDescription = metaDesc != null ? metaDesc.attr("content") : "";

        // Headings
        data.h1s = texts(doc.select("h1"));
        data.h2s = texts(doc.select("h2"));

        // Images
        Elements images = doc.select("img");
        data.imageCount = images.size();
        data.imagesWithMissingAlt = new ArrayList<>();
        for (Element img : images) {
            if (img.attr("alt").trim().isEmpty()) {
                String src = img.attr("abs:src");
                if (src.isBlank()) src = img.attr("src");
                if (!src.isBlank()) data.imagesWithMissingAlt.add(src);
            }
        }
        data.missingAltCount = data.imagesWithMissingAlt.size();

        // Canonical
        Element canonical = doc.selectFirst("link[rel=canonical]");
        data.canonicalUrl = canonical != null ? canonical.attr("href") : "";

        // Internal and external links
        int internalCount = 0, externalCount = 0;
        for (Element a : doc.select("a[href]")) {
            String href = a.attr("abs:href");
            if (href.isBlank() || href.startsWith("#") ||
                href.startsWith("mailto:") || href.startsWith("tel:") ||
                href.startsWith("javascript:")) continue;
            if (href.contains(data.baseDomain)) internalCount++;
            else externalCount++;
        }
        data.internalLinkCount = internalCount;
        data.externalLinkCount = externalCount;

        // Schema types from JSON-LD
        data.schemaTypes = extractSchemaTypes(doc);

        // Schema field counts (must extract before script removal)
        int[] schemaCounts = extractSchemaFieldCounts(doc);
        data.validSchemaFields = schemaCounts[0];
        data.totalSchemaFields = schemaCounts[1];

        // Author - check common meta patterns
        data.authorName = extractAuthor(doc);

        // Dates
        data.datePublished = extractMeta(doc, "article:published_time", "datePublished");
        data.dateModified = extractMeta(doc, "article:modified_time", "dateModified");
        data.daysSinceLastUpdate = computeDaysSinceUpdate(data.dateModified, data.datePublished);

        // Body text — remove nav, footer, script, style first
        doc.select("nav, footer, script, style, header").remove();

        // Heading and list counts (nav/header/footer already removed)
        data.headingCount = doc.select("h1, h2, h3, h4, h5, h6").size();
        data.listCount    = doc.select("ul, ol").size();

        String bodyText = doc.select("main, article, .content, #content, body").text();

        // Quotations: blockquotes in doc + inline "..." patterns in body text
        data.quotationCount = doc.select("blockquote").size() + countInlineQuotes(bodyText);

        // Full body text metrics (before truncation)
        data.totalWords            = countMeaningfulWords(bodyText);
        data.wordCount             = data.totalWords;
        data.totalSentences        = countSentences(bodyText);
        data.factualStatementCount = countFactualStatements(bodyText);
        data.avgChunkSize          = data.headingCount > 0
                                     ? (double) data.totalWords / data.headingCount
                                     : data.totalWords;
        data.technicalTermCount    = countTechnicalTerms(bodyText);

        data.bodyText = bodyText.length() > 2000 ? bodyText.substring(0, 2000) : bodyText;

        log.info("Crawled: " + url + " | words: " + data.wordCount + " | schema: " + data.schemaTypes);
        return data;
    }

    private List<String> texts(Elements elements) {
        List<String> result = new ArrayList<>();
        for (Element el : elements) {
            String t = el.text().trim();
            if (!t.isEmpty()) result.add(t);
        }
        return result;
    }

    private List<String> extractSchemaTypes(Document doc) {
        List<String> types = new ArrayList<>();
        Pattern typePattern = Pattern.compile("\"@type\"\\s*:\\s*\"([^\"]+)\"");
        for (Element script : doc.select("script[type=application/ld+json]")) {
            Matcher m = typePattern.matcher(script.html());
            while (m.find()) {
                String type = m.group(1);
                if (!types.contains(type)) types.add(type);
            }
        }
        return types;
    }

    private String extractAuthor(Document doc) {
        // Try meta tags
        String[] authorAttrs = {
            "author", "article:author", "og:article:author",
            "twitter:creator", "DC.creator"
        };
        for (String attr : authorAttrs) {
            Element el = doc.selectFirst("meta[name=" + attr + "]");
            if (el == null) el = doc.selectFirst("meta[property=" + attr + "]");
            if (el != null && !el.attr("content").trim().isEmpty()) {
                return el.attr("content").trim();
            }
        }

        // Try JSON-LD
        Pattern authorPattern = Pattern.compile("\"author\"\\s*:\\s*\\{[^}]*\"name\"\\s*:\\s*\"([^\"]+)\"");
        for (Element script : doc.select("script[type=application/ld+json]")) {
            Matcher m = authorPattern.matcher(script.html());
            if (m.find()) return m.group(1);
        }

        // Try rel=author or common class names
        Element authorEl = doc.selectFirst("[rel=author], .author-name, .byline, [itemprop=author]");
        if (authorEl != null && !authorEl.text().trim().isEmpty()) {
            return authorEl.text().trim();
        }

        return "";
    }

    private int[] extractSchemaFieldCounts(Document doc) {
        int validFields = 0;
        int totalExpected = 0;
        for (Element script : doc.select("script[type=application/ld+json]")) {
            try {
                JsonNode root = schemaMapper.readTree(script.html());
                List<JsonNode> schemas = new ArrayList<>();
                if (root.has("@graph") && root.get("@graph").isArray()) {
                    root.get("@graph").forEach(schemas::add);
                } else {
                    schemas.add(root);
                }
                for (JsonNode schema : schemas) {
                    String type = schema.has("@type") ? schema.get("@type").asText() : "";
                    totalExpected += SCHEMA_EXPECTED_FIELDS.getOrDefault(type, 5);
                    validFields += countNonEmptySchemaFields(schema, new HashSet<>());
                }
            } catch (Exception ignored) {}
        }
        return new int[]{ validFields, totalExpected };
    }

    private int countNonEmptySchemaFields(JsonNode node, Set<String> counted) {
        int count = 0;
        Iterator<Map.Entry<String, JsonNode>> fields = node.fields();
        while (fields.hasNext()) {
            Map.Entry<String, JsonNode> entry = fields.next();
            String key = entry.getKey();
            JsonNode val = entry.getValue();
            if (key.startsWith("@")) continue;
            if (counted.contains(key)) continue;
            if (val.isNull() || (val.isTextual() && val.asText().trim().isEmpty())) continue;
            counted.add(key);
            count++;
            if (val.isObject()) count += countNonEmptySchemaFields(val, counted);
        }
        return count;
    }

    private Integer computeDaysSinceUpdate(String dateModified, String datePublished) {
        String dateStr = (dateModified != null && !dateModified.isBlank()) ? dateModified : datePublished;
        if (dateStr == null || dateStr.isBlank() || dateStr.length() < 10) return null;
        try {
            LocalDate date = LocalDate.parse(dateStr.substring(0, 10));
            return (int) ChronoUnit.DAYS.between(date, LocalDate.now());
        } catch (Exception e) {
            return null;
        }
    }

    private int countInlineQuotes(String text) {
        if (text == null || text.isBlank()) return 0;
        Matcher m = Pattern.compile("\"[^\"]{3,}\"").matcher(text);
        int count = 0;
        while (m.find()) count++;
        return count;
    }

    private int countMeaningfulWords(String text) {
        if (text == null || text.isBlank()) return 0;
        int count = 0;
        for (String token : text.split("\\s+")) {
            if (!token.replaceAll("[^a-zA-Z]", "").isEmpty()) count++;
        }
        return count;
    }

    private int countSentences(String text) {
        if (text == null || text.isBlank()) return 0;
        int count = 0;
        for (String s : text.split("[.?!]+")) {
            if (s.trim().split("\\s+").length >= 3) count++;
        }
        return count;
    }

    private int countFactualStatements(String text) {
        if (text == null || text.isBlank()) return 0;
        int count = 0;
        for (String s : text.split("[.?!]+")) {
            String trimmed = s.trim();
            if (trimmed.split("\\s+").length < 3) continue;
            if (FACTUAL_PATTERN.matcher(trimmed).find()) count++;
        }
        return count;
    }

    private int countTechnicalTerms(String text) {
        if (text == null || text.isBlank()) return 0;
        Matcher m = TECHNICAL_TERM_PATTERN.matcher(text);
        Set<String> seen = new HashSet<>();
        while (m.find()) {
            String term = m.group().trim();
            if (term.length() > 3) seen.add(term.toLowerCase());
        }
        return seen.size();
    }

    private String extractMeta(Document doc, String... names) {
        for (String name : names) {
            Element el = doc.selectFirst("meta[property=" + name + "]");
            if (el == null) el = doc.selectFirst("meta[name=" + name + "]");
            if (el == null) el = doc.selectFirst("time[itemprop=" + name + "]");
            if (el != null) {
                String val = el.attr("content");
                if (val.isEmpty()) val = el.attr("datetime");
                if (!val.isEmpty()) return val;
            }
        }
        // Try JSON-LD
        String key = names[names.length - 1];
        Pattern p = Pattern.compile("\"" + key + "\"\\s*:\\s*\"([^\"]+)\"");
        for (Element script : doc.select("script[type=application/ld+json]")) {
            Matcher m = p.matcher(script.html());
            if (m.find()) return m.group(1);
        }
        return "";
    }
}

File: src/services/crawlerService.js
const axios = require('axios');
const cheerio = require('cheerio');
const config = require('../config');

const SCHEMA_EXPECTED_FIELDS = {
  Article: 8, BlogPosting: 8, NewsArticle: 8,
  Product: 9, FAQPage: 4, WebSite: 5,
  WebPage: 5, Organization: 6, LocalBusiness: 7,
  Person: 5,
};

const FACTUAL_PATTERN = /\d+([,.]\d+)?(%|\s*(percent|million|billion|thousand))?|\b(according to|research|study|studies|report|survey|data|statistics|found that)\b/gi;
const TECHNICAL_TERM_PATTERN = /\b[A-Z][a-zA-Z0-9]*(?:[\s-][A-Z][a-zA-Z0-9]*){0,2}\b/g;

async function crawl(url) {
  const parsedUrl = new URL(url);
  const baseDomain = parsedUrl.hostname;
  const slug = parsedUrl.pathname;

  const response = await axios.get(url, {
    timeout: config.scanner.connectTimeoutMs,
    headers: {
      'User-Agent': config.scanner.userAgent,
      'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8',
      'Accept-Language': 'en-US,en;q=0.9',
      'Accept-Encoding': 'gzip, deflate, br',
      'Cache-Control': 'no-cache',
      'Pragma': 'no-cache',
      'Upgrade-Insecure-Requests': '1',
      'Sec-Fetch-Dest': 'document',
      'Sec-Fetch-Mode': 'navigate',
      'Sec-Fetch-Site': 'none',
      'Sec-Fetch-User': '?1',
    },
    maxRedirects: 5,
  });

  const $ = cheerio.load(response.data);

  // Title
  const title = $('title').first().text().trim();

  // Meta description
  const metaDesc = $('meta[name="description"]').attr('content') ||
    $('meta[property="og:description"]').attr('content') || '';

  // Headings
  const h1s = $('h1').map((_, el) => $(el).text().trim()).get().filter(Boolean);
  const h2s = $('h2').map((_, el) => $(el).text().trim()).get().filter(Boolean);

  // Images
  const images = $('img');
  const imageCount = images.length;
  const imagesWithMissingAlt = [];
  images.each((_, el) => {
    const alt = $(el).attr('alt') || '';
    if (!alt.trim()) {
      const src = $(el).attr('src') || '';
      if (src) imagesWithMissingAlt.push(src);
    }
  });
  const missingAltCount = imagesWithMissingAlt.length;

  // Canonical
  const canonicalUrl = $('link[rel="canonical"]').attr('href') || '';

  // Internal and external links
  let internalLinkCount = 0, externalLinkCount = 0;
  $('a[href]').each((_, el) => {
    const href = $(el).attr('href') || '';
    if (!href || href.startsWith('#') || href.startsWith('mailto:') ||
        href.startsWith('tel:') || href.startsWith('javascript:')) return;
    try {
      const absolute = href.startsWith('http') ? href : new URL(href, url).href;
      const linkDomain = new URL(absolute).hostname;
      if (linkDomain === baseDomain) internalLinkCount++;
      else externalLinkCount++;
    } catch {
      // skip malformed
    }
  });

  // Schema types
  const schemaTypes = extractSchemaTypes($);

  // Schema field counts
  const { validSchemaFields, totalSchemaFields } = extractSchemaFieldCounts($);

  // Author
  const authorName = extractAuthor($);

  // Dates
  const datePublished = extractMeta($, ['article:published_time', 'datePublished']);
  const dateModified = extractMeta($, ['article:modified_time', 'dateModified']);
  const daysSinceLastUpdate = computeDaysSinceUpdate(dateModified, datePublished);

  // Remove nav/header/footer/script/style before body text extraction
  $('nav, footer, script, style, header').remove();

  const headingCount = $('h1, h2, h3, h4, h5, h6').length;
  const listCount = $('ul, ol').length;

  const bodyText = $('main, article, .content, #content, body').first().text() ||
    $('body').text();

  const quotationCount = $('blockquote').length + countInlineQuotes(bodyText);

  const totalWords = countMeaningfulWords(bodyText);
  const wordCount = totalWords;
  const totalSentences = countSentences(bodyText);
  const factualStatementCount = countFactualStatements(bodyText);
  const avgChunkSize = headingCount > 0 ? totalWords / headingCount : totalWords;
  const technicalTermCount = countTechnicalTerms(bodyText);

  console.log(`Crawled: ${url} | words: ${wordCount} | schema: ${JSON.stringify(schemaTypes)}`);

  return {
    url, slug, title, metaDescription: metaDesc, h1s, h2s,
    imageCount, missingAltCount, imagesWithMissingAlt,
    canonicalUrl, internalLinkCount, externalLinkCount,
    schemaTypes, validSchemaFields, totalSchemaFields,
    authorName, datePublished, dateModified, daysSinceLastUpdate,
    wordCount, totalWords, totalSentences, factualStatementCount,
    quotationCount, headingCount, listCount, avgChunkSize, technicalTermCount,
    baseDomain,
    bodyText: bodyText.length > 2000 ? bodyText.substring(0, 2000) : bodyText,
  };
}

function extractSchemaTypes($) {
  const types = [];
  const typePattern = /"@type"\s*:\s*"([^"]+)"/g;
  $('script[type="application/ld+json"]').each((_, el) => {
    const html = $(el).html() || '';
    let m;
    while ((m = typePattern.exec(html)) !== null) {
      if (!types.includes(m[1])) types.push(m[1]);
    }
    typePattern.lastIndex = 0;
  });
  return types;
}

function extractSchemaFieldCounts($) {
  let validSchemaFields = 0, totalSchemaFields = 0;
  $('script[type="application/ld+json"]').each((_, el) => {
    try {
      const root = JSON.parse($(el).html() || '{}');
      const schemas = root['@graph'] && Array.isArray(root['@graph'])
        ? root['@graph'] : [root];
      for (const schema of schemas) {
        const type = schema['@type'] || '';
        totalSchemaFields += SCHEMA_EXPECTED_FIELDS[type] || 5;
        validSchemaFields += countNonEmptySchemaFields(schema, new Set());
      }
    } catch { /* ignore invalid JSON */ }
  });
  return { validSchemaFields, totalSchemaFields };
}

function countNonEmptySchemaFields(node, counted) {
  if (typeof node !== 'object' || node === null) return 0;
  let count = 0;
  for (const [key, val] of Object.entries(node)) {
    if (key.startsWith('@')) continue;
    if (counted.has(key)) continue;
    if (val === null || val === undefined || val === '') continue;
    counted.add(key);
    count++;
    if (typeof val === 'object') count += countNonEmptySchemaFields(val, counted);
  }
  return count;
}

function extractAuthor($) {
  const authorAttrs = ['author', 'article:author', 'og:article:author', 'twitter:creator', 'DC.creator'];
  for (const attr of authorAttrs) {
    const content = $(`meta[name="${attr}"]`).attr('content') ||
                    $(`meta[property="${attr}"]`).attr('content');
    if (content && content.trim()) return content.trim();
  }

  // Try JSON-LD
  const authorPattern = /"author"\s*:\s*\{[^}]*"name"\s*:\s*"([^"]+)"/;
  let authorFound = '';
  $('script[type="application/ld+json"]').each((_, el) => {
    if (authorFound) return;
    const m = authorPattern.exec($(el).html() || '');
    if (m) authorFound = m[1];
  });
  if (authorFound) return authorFound;

  // Try selectors
  const authorEl = $('[rel="author"], .author-name, .byline, [itemprop="author"]').first();
  if (authorEl.length) {
    const text = authorEl.text().trim();
    if (text) return text;
  }

  return '';
}

function extractMeta($, names) {
  for (const name of names) {
    const el = $(`meta[property="${name}"]`).first() ||
               $(`meta[name="${name}"]`).first();
    const content = el.attr('content') || '';
    if (content) return content;

    const timeEl = $(`time[itemprop="${name}"]`).first();
    const datetime = timeEl.attr('datetime') || timeEl.attr('content') || '';
    if (datetime) return datetime;
  }
  // Try JSON-LD
  const key = names[names.length - 1];
  const p = new RegExp(`"${key}"\\s*:\\s*"([^"]+)"`);
  let found = '';
  $('script[type="application/ld+json"]').each((_, el) => {
    if (found) return;
    const m = p.exec($(el).html() || '');
    if (m) found = m[1];
  });
  return found;
}

function computeDaysSinceUpdate(dateModified, datePublished) {
  const dateStr = (dateModified && dateModified.trim()) ? dateModified : datePublished;
  if (!dateStr || dateStr.length < 10) return null;
  try {
    const date = new Date(dateStr.substring(0, 10));
    if (isNaN(date.getTime())) return null;
    const now = new Date();
    return Math.floor((now - date) / (1000 * 60 * 60 * 24));
  } catch {
    return null;
  }
}

function countInlineQuotes(text) {
  if (!text) return 0;
  const matches = text.match(/"[^"]{3,}"/g);
  return matches ? matches.length : 0;
}

function countMeaningfulWords(text) {
  if (!text) return 0;
  return text.split(/\s+/).filter(t => t.replace(/[^a-zA-Z]/g, '').length > 0).length;
}

function countSentences(text) {
  if (!text) return 0;
  return text.split(/[.?!]+/).filter(s => s.trim().split(/\s+/).length >= 3).length;
}

function countFactualStatements(text) {
  if (!text) return 0;
  let count = 0;
  for (const s of text.split(/[.?!]+/)) {
    const trimmed = s.trim();
    if (trimmed.split(/\s+/).length < 3) continue;
    FACTUAL_PATTERN.lastIndex = 0;
    if (FACTUAL_PATTERN.test(trimmed)) count++;
  }
  return count;
}

function countTechnicalTerms(text) {
  if (!text) return 0;
  const seen = new Set();
  let m;
  const re = new RegExp(TECHNICAL_TERM_PATTERN.source, 'g');
  while ((m = re.exec(text)) !== null) {
    const term = m[0].trim();
    if (term.length > 3) seen.add(term.toLowerCase());
  }
  return seen.size;
}

module.exports = { crawl };
